### PR TITLE
feat: vault secrets

### PR DESCRIPTION
**Pull request by ovasdi**
https://artsyproduct.atlassian.net/browse/PHIRE-803

This enables Vault secrets for this projects deployments following the [playbook](https://www.notion.so/artsy/Migration-of-Artsy-Projects-Secrets-07e5ce9e1ed044dba8d4b7c3f0ae1ccf) which documents the steps.

### Unleash images

This project is currently setup with two deployments: `artsy-unleash` and `artsy-unleash-proxy`.
The former, `artsy-unleash` deployment, is using an artsy managed [Docker image](https://github.com/artsy/artsy-unleash/blob/main/Dockerfile) which currently introduces only minimal configuration of port for running the [unleash-server](https://github.com/Unleash/unleash) ([docker-hub](https://hub.docker.com/r/unleashorg/unleash-server)). The other deployment, [`artsy-unleash-proxy`](https://github.com/Unleash/unleash-proxy) ([docker-hub](https://hub.docker.com/r/unleashorg/unleash-proxy)), uses the standard _unleash managed_ [docker image](https://github.com/artsy/artsy-unleash/blob/e2da647a3357811b4d53127ed35f00b9a1cbeeea/hokusai/production.yml#L178).

EDIT: After receiving feedback, to continue to use proxy image managed by unleash, this edits the spec to allow for mounting and execution of the `load_secrets_and_run.sh` without having to manage our own docker image.

<details><summary>Outdated description</summary>
<p>

Looking at the environment variables, there are currently two sensitive vars that should live in Vault: `DATABASE_URL` and `UNLEASH_API_TOKEN`. The latter appears to be set in both: `artsy-unleash-environment` and `artsy-unleash-proxy-environment` while the former only in `artsy-unleash-environment`. `UNLEASH_API_TOKEN` _may_ need to be accessible by both deployments but by my understanding, it is required by the proxy, which means `artsy-unleash-proxy` image must be managed by artsy and modified to start with `load_secrets_and_run.sh` script.

### Start managing artsy-unleash-proxy

In addition to Vault specific changes, this also introduces the management of unleash proxy so that it is possible to use Vault secrets with that deployment. A separate `Dockerfile.proxy` is added and also a script to allow to build and push the image.

### Entrypoint

After some digging it turns out that the Entrypoint the unleash-server (and proxy) is using is located in `/usr/local/bin/`. See [server image layers](https://hub.docker.com/layers/unleashorg/unleash-server/latest/images/sha256-7624eb900a2e62460a416fdbd5adc29c7db06fda50830faf302e0c659d08a889?context=explore) for more details, layer 7 specifically for the path to entrypoint.

The contents of entrypoint script are:
```
➜  artsy-unleash git:(ovasdi/vault) ✗ docker run -it unleashorg/unleash-server:latest cat /usr/local/bin/docker-entrypoint.sh
#!/bin/sh
set -e

# Run command with node if the first argument contains a "-" or is not a system command. The last
# part inside the "{}" is a workaround for the following bug in ash/dash:
# https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=874264
if [ "${1#-}" != "${1}" ] || [ -z "$(command -v "${1}")" ] || { [ -f "${1}" ] && ! [ -x "${1}" ]; }; then
  set -- node "$@"
fi

exec "$@"
```

So it appears that the command should just get passed through the load_script and into this entrypoint which executes the command.

TODO:
- [ ] create (terraform) artsy-unleash-proxy ecr repo (if this is how we want to do it)
- [ ] build and push the artsy-unleash-proxy image 

</p>
</details>

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,3 +2,9 @@ FROM unleashorg/unleash-server
 
 ENV HTTP_PORT 8080
 EXPOSE 8080
+
+COPY ./scripts/load_secrets_and_run.sh load_secrets_and_run.sh
+
+ENTRYPOINT ["/usr/local/bin/docker-entrypoint.sh", "./load_secrets_and_run.sh"]
+
+CMD ["node", "index.js"]
```
Comment (ovasdi) [inline Dockerfile:10]: After adding the ENTRYPOINT I tried starting the docker image but it wouldnt actually start the server so I think this is needed.

File: Dockerfile.proxy [hunk 2]
```diff
@@ -0,0 +1,10 @@
+FROM unleashorg/unleash-proxy
+
+ENV PORT 4242
+EXPOSE 4242
+
+COPY ./scripts/load_secrets_and_run.sh load_secrets_and_run.sh
+
+ENTRYPOINT ["/usr/local/bin/docker-entrypoint.sh", "./load_secrets_and_run.sh"]
+
+CMD ["/bin/sh" "-c" "./server.sh"]
```
Comment (ovasdi) [inline Dockerfile.proxy:10]: Got this CMD from the [image layers](https://hub.docker.com/layers/unleashorg/unleash-proxy/latest/images/sha256-748fc3ae83ed3883d5b53ef21ab943f5c7c54291821aaf0fe61e5a2f835df7fd?context=explore)

File: hokusai/production.yml [hunk 3]
```diff
@@ -173,9 +215,15 @@ spec:
...
+        - configMapRef:
+            name: secrets-config
         - configMapRef:
             name: artsy-unleash-proxy-environment
-        image: unleashorg/unleash-proxy
+        volumeMounts:
+        - name: secrets
+          mountPath: /secrets
+          readOnly: true
+        image: 585031190124.dkr.ecr.us-east-1.amazonaws.com/artsy-unleash-proxy:latest
```
Comment (ovasdi) [inline hokusai/production.yml:226]: Looking for some advice on whether this is the way we would like to go. I'd prefer to use the [artsy-unleash](https://us-east-1.console.aws.amazon.com/ecr/repositories/private/585031190124/artsy-unleash?region=us-east-1) ecr repo but im not sure if thats going to complicate things. Maybe its more straight forward to do this?

File: scripts/build_push_proxy.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Comment (ovasdi) [inline scripts/build_push_proxy.sh:1]: Maybe this script is not necessary at all? What do the reviewers think?

**Comment by ovasdi** [comment]
@joeyAghion perhaps you could confirm or deny the bit about `UNLEASH_API_TOKEN` being required by the proxy deployment?

**Comment by joeyAghion** [comment]
Yes [the proxy docs mention](https://docs.getunleash.io/reference/unleash-proxy#configuration-options) that it needs an API token to start.

Does the secrets-loading approach _require_ that we customize every Dockerfile and host every docker image? That's a shame when otherwise we'd be able to use public images, such as for services like these. Can the loading be done by passing a custom `ENTRYPOINT`/`command` in the spec instead?

**Comment by ovasdi** [comment]
> Yes [the proxy docs mention](https://docs.getunleash.io/reference/unleash-proxy#configuration-options) that it needs an API token to start.
> 
> Does the secrets-loading approach _require_ that we customize every Dockerfile and host every docker image? That's a shame when otherwise we'd be able to use public images, such as for services like these. Can the loading be done by passing a custom `ENTRYPOINT`/`command` in the spec instead?

@joeyAghion Thanks for confirming.
Don't know if its possible to use the spec that way but appreciate the question you posed and will see if its possible.

**Comment by ovasdi** [comment]
> > Yes [the proxy docs mention](https://docs.getunleash.io/reference/unleash-proxy#configuration-options) that it needs an API token to start.
> > Does the secrets-loading approach _require_ that we customize every Dockerfile and host every docker image? That's a shame when otherwise we'd be able to use public images, such as for services like these. Can the loading be done by passing a custom `ENTRYPOINT`/`command` in the spec instead?
> 
> @joeyAghion Thanks for confirming. Don't know if its possible to use the spec that way but appreciate the question you posed and will see if its possible.

Using the entrypoint via spec command seems viable but the hurdle is the [script](https://github.com/artsy/artsy-unleash/pull/15/files#diff-83c0f7c57dabdc111cbbea7cc5c284c56ee7f61c3e35e25d5ba8c1c49d474e68) that needs to be executed. I wonder if this could be done by creating a ConfigMap (or secret) to include the file in the deployment. Then mount the config map as a volume and set the command to execute the script 🤔

File: hokusai/production.yml (hunk 3 above)
Comment (artsyjian) [reply hokusai/production.yml:226]: i think Hokusai currently maps a Github repo with an ECR repo, one to one. so we would have to use a separate ECR repo.

the problem then becomes, how do we get CI to build two Docker images.

Comment (artsyjian) [reply hokusai/production.yml:226]: oh i see, the suggestion is to build proxy image [locally](scripts/build_push_proxy.sh).

File: scripts/build_push_proxy.sh (hunk 4 above)
Comment (artsyjian) [reply scripts/build_push_proxy.sh:1]: nice script! yeah, CI/Hokusai will just build the canonical image (`artsy-unleash`), so using this script to locally build the proxy image, is a solution.

**Review by artsyjian — APPROVED** [review]
Amazing! Thanks for working through so many challenges presented by this repo.

File: hokusai/production.yml (hunk 3 above)
Comment (ovasdi) [reply hokusai/production.yml:226]: > the problem then becomes, how do we get CI to build two Docker images.

Right, I couldnt think of one or find an example of us already doing that so reached for the manual script...

**Comment by ovasdi** [comment]
OK I made an attempt to do what @joeyAghion asked about in https://github.com/artsy/artsy-unleash/pull/15/commits/89266042c6edeab031c8a4b3f305e6a412c3f7d1

- go back to using `image: unleashorg/unleash-proxy` and remove the proxy dockerization
- adds ConfigMap with `load_secrets_and_run.sh` script and mounts it
- adds an additional step to the `initContainer` to set the executable permissions for the script
- adds `command` and `args` to main container to define an entrypoint and and pass the start command

Im wondering if it would be OK to test this out without merging the PR.
Maybe I can apply the whole `staging.yml` spec from my local and see if it sticks, if the container starts successfully and proxy initializes without issues. 🤔 

Something like:
```
cd hokusai/
kubectl --context staging apply -f staging.yml -n default
```

It would be great if this work and if not I can always re-apply the staging spec in `main`  then nuke the last commit and go back to doing this the dockerized way.

If I cant make it work:
```
git checkout main
kubectl --context staging apply -f staging.yml -n default
```

What do you think @artsy/platform and specifically @artsyjian? Am I missing something critical that makes this a no go?

**Comment by artsyjian** [comment]
yeah, for just spec changes, you can apply them with `staging update ...`.

**Comment by ovasdi** [comment]
> yeah, for just spec changes, you can apply them with `staging update ...`.

Oh right 🤦‍♂️, we can just use hokusai commands to do that.

**Comment by artsyjian** [comment]
> use hokusai commands

oh yes, `hokusai staging update ...`. i had read your `kubectl` as `hokusai` so didn't specify. ha.

**Comment by ovasdi** [comment]
HA
```
➜  artsy-unleash git:(ovasdi/vault) hokusai staging update
ERROR: Not on main branch!  Aborting
```

**Comment by artsyjian** [comment]
> HA
> 
> ```
> ➜  artsy-unleash git:(ovasdi/vault) hokusai staging update
> ERROR: Not on main branch!  Aborting
> ```

`--skip-checks`

**Comment by ovasdi** [comment]
It appears the configmap was created and the proxy deployment configured but for some reason the proxy pod didnt get restarted
```sh
hokusai staging update --skip-checks
deployment.apps/artsy-unleash-web configured
horizontalpodautoscaler.autoscaling/artsy-unleash-web unchanged
service/artsy-unleash-web-internal unchanged
ingress.networking.k8s.io/artsy-unleash unchanged
deployment.apps/artsy-unleash-proxy-web configured
configmap/unleash-proxy-scripts created
horizontalpodautoscaler.autoscaling/artsy-unleash-proxy-web unchanged
service/artsy-unleash-proxy-web-internal unchanged
ingress.networking.k8s.io/artsy-unleash-proxy unchanged
Updated Kubernetes environment /Users/ovasdi/code/artsy-unleash/hokusai/staging.yml
```

```sh
➜  ~ kubectl --context staging get pods | grep unleash
artsy-unleash-proxy-web-5dc89f44-qq5vl                            1/1     Running                  0               23d
artsy-unleash-web-7cc4f6bfc-224r9                                 1/1     Running                  0               4m34s
```

**Comment by ovasdi** [comment]
I also noticed similarly that when i ran `hokusai staging refresh` earlier it didnt affect the proxy pod, only the `artsy-unleash-web` pod restarted.

**Comment by artsyjian** [comment]
> I also noticed similarly that when i ran `hokusai staging refresh` earlier it didnt affect the proxy pod, only the `artsy-unleash-web` pod restarted.

[this](https://github.com/artsy/artsy-unleash/pull/15#issuecomment-2492470080) may fix it. hokusai [expects](https://github.com/artsy/hokusai/blob/d60e552723e43ee44a5e09ba1ae2e35c7cbc008b/hokusai/commands/kubernetes.py#L148) `app` label to match project name. in this case, the project name is `artsy-unleash` while proxy's `app` label is [artsy-unleash-proxy](https://github.com/artsy/artsy-unleash/blob/e2da647a3357811b4d53127ed35f00b9a1cbeeea/hokusai/staging.yml#L142). by that commit, i suggest changing the label to conform, and use `component` label to distinguish proxy deployment/service.

**Comment by ovasdi** [comment]
@artsyjian I was reviewing the spec and noticed that some labels were still out of date, specifically under _selector_ and _template_ keys. Its essentially the same change you made but in [3 more locations](https://github.com/artsy/artsy-unleash/pull/15/commits/9f119c12542a6a17660b6bd3c9da31fb4c20bda6) per environment. I think this is more correct now, would you agree?

File: hokusai/production.yml [hunk 5]
```diff
@@ -200,11 +258,14 @@ spec:
...
               - -c
               - sleep 5
+        command: ["/usr/local/bin/docker-entrypoint.sh", "/usr/local/bin/load_secrets_and_run.sh"]
+        args: ["/bin/sh", "-c", "./server.sh"]
       dnsPolicy: ClusterFirst
       dnsConfig:
         options:
         - name: ndots
           value: '1'
+      serviceAccountName: artsy-unleash-proxy
```
Comment (artsyjian) [inline hokusai/production.yml:268]: i planned to just use [one service account](https://github.com/artsy/substance/blob/98606010adefa882a227f986cae29d3b5dd13263/services/service-accounts-k8s-apps.yml#L32) on k8s side. can just use `artsy-unleash` here.

File: hokusai/staging.yml [hunk 6]
```diff
@@ -200,11 +258,14 @@ spec:
...
               - -c
               - sleep 5
+        command: ["/usr/local/bin/docker-entrypoint.sh", "/usr/local/bin/load_secrets_and_run.sh"]
+        args: ["/bin/sh", "-c", "./server.sh"]
       dnsPolicy: ClusterFirst
       dnsConfig:
         options:
         - name: ndots
           value: '1'
+      serviceAccountName: artsy-unleash-proxy
```
Comment (artsyjian) [inline hokusai/staging.yml:268]: same as comment for prod, `artsy-unleash` may be fine.

File: hokusai/production.yml (hunk 5 above)
Comment (artsyjian) [reply hokusai/production.yml:268]: ouch, i am now seeing that there are two configmaps in this project. and they conflict in var values, e.g. `UNLEASH_URL`.

Comment (artsyjian) [reply hokusai/production.yml:268]: that means Vault has to mirror that. we have to have two prefixes, one per deployment:

```
kvv2/data/kubernetes/apps/artsy-unleash/*
kvv2/data/kubernetes/apps/artsy-unleash-proxy/*
```

the first one already [exists](https://github.com/artsy/infrastructure/blob/8466051372ced68ba26ff00d8e835391efc7e95d/terraform/staging/vault/k8s_apps_role_policy.tf#L22). we have to add the 2nd one.

Comment (artsyjian) [reply hokusai/production.yml:268]: so yeah, may as well add a `artsy-unleash-proxy` service account, too.

File: hokusai/staging.yml (hunk 6 above)
Comment (artsyjian) [reply hokusai/staging.yml:268]: nope. never mind. updated that comment.

**Comment by artsyjian** [comment]
> @artsyjian I was reviewing the spec and noticed that some labels were still out of date, specifically under _selector_ and _template_ keys. Its essentially the same change you made but in [3 more locations](https://github.com/artsy/artsy-unleash/pull/15/commits/9f119c12542a6a17660b6bd3c9da31fb4c20bda6) per environment. I think this is more correct now, would you agree?

good catch! yeah, Service has to be updated. looks good!

File: hokusai/production.yml (hunk 5 above)
Comment (ovasdi) [reply hokusai/production.yml:261]: Created service account -- https://github.com/artsy/substance/pull/402
Created aws role and policy -- https://github.com/artsy/infrastructure/pull/746
Created a path in vault 
```
vault kv put -mount kvv2 -cas=0 kubernetes/apps/artsy-unleash-proxy/PLACEHOLDER PLACEHOLDER=placeholder
```

**Comment by ovasdi** [comment]
It appears the approach using a mounted script and a `command` with `args` worked. After updating the deployment locally I see the following output [get logged](https://my.papertrailapp.com/groups/3674473/events?q=program%3Aunleash&focus=1798447444569305089&selected=1798447444569305089):
```sh
SECRETS_FILE env var is defined. Sourcing secrets file...
Running command: /bin/sh -c ./server.sh
Unleash-proxy is listening on port 4242!
```

```
➜  ~ curl --request GET --url https://unleashprx-staging.artsy.net/proxy/health
ok
➜  kubectl --context staging get pods | grep unleash
artsy-unleash-proxy-web-66877bbf4f-pnvng                          1/1     Running                  0               37m
artsy-unleash-web-74759d57dc-lktqj                                1/1     Running                  0               40m
```

---

Question @artsyjian:

Seeing as this project uses two configmaps, I am trying to understand if the opstools script can be used for migrating _both_ configmaps?
I am looking at the [migrate.py](https://github.com/artsy/opstools/blob/fc86ee46c59634e4741a4fb036644ec9ba74e6ac/src/migrate_config_secrets/migrate_config_secrets/migrate.py#L57) and it looks like the `artsy_project` command arg is used for both, [targeting the configmap](https://github.com/artsy/opstools/blob/fc86ee46c59634e4741a4fb036644ec9ba74e6ac/src/migrate_config_secrets/migrate_config_secrets/migrate.py#L57) and setting the [project repository directory](https://github.com/artsy/opstools/blob/fc86ee46c59634e4741a4fb036644ec9ba74e6ac/src/migrate_config_secrets/migrate_config_secrets/migrate.py#L89C51-L89C64). I was thinking that maybe it would work if I ran the script twice, once with `artsy-unleash` and second time with `artsy-unleash-proxy` but for the second run, I copy the repo into a separate directory named `artsy-unleash-proxy`. Would that work then to move secrets across both configmaps using the migrate script?

Or should I just migrate the `artsy-unleash-proxy` secrets manually?

**Comment by artsyjian** [comment]
> Seeing as this project uses two configmaps, I am trying to understand if the opstools script can be used for migrating _both_ configmaps?

the script uses Hokusai env commands which i think only sees the canonical configmap (`artsy-unleash-environment`). so yeah, probably have to migrate manually.

**Pull request merged**